rfd/0000-prototype-job-worker.md

```diff
@@ -0,0 +1,203 @@
+namespaces. The following code snippet demonstrates how a new process can be
+started in a new set of namespaces:
+
+```go
+import "syscall"
+
+func setupNamespaces() error {
+    // Unshare creates new namespaces for the calling process
+    if err := syscall.Unshare(syscall.CLONE_NEWPID | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS); err != nil {
+        return fmt.Errorf("failed to create namespaces: %v", err)
+    }
+    return nil
+}
+```
+
+#### Cgroup Implementation
+Control groups (cgroups) are another kernel feature that limits, accounts for,
+and isolates the resource usage (CPU, memory, disk I/O, etc.) of process
+groups. We will implement cgroups to ensure that each job runs within
+predefined resource limits, improving system stability and preventing any job
+from monopolizing system resources.
+
+##### Configuration Approach
+Cgroups will be managed by interacting with the cgroup filesystem typically
+mounted under /sys/fs/cgroup. Here is how we plan to create and manage cgroups:
+
+```go
+import (
+    "io/ioutil"
+    "path/filepath"
+)
+
+func createCgroup(path, resource, limit string) error {
+    cgroupPath := filepath.Join("/sys/fs/cgroup", resource, path)
+    if err := os.MkdirAll(cgroupPath, 0755); err != nil {
+        return err
+    }
+    return ioutil.WriteFile(filepath.Join(cgroupPath, "tasks"), []byte(strconv.Itoa(os.Getpid())), 0644)
+}
+```
+
+### Command Usage
+
+```
+Usage: jobworker [command] [options]
+
+A CLI tool to interact with the Job Worker Service, providing secure job management capabilities through mTLS.
+
+Commands:
+  start       Start a new job with the specified command and arguments.
+  stop        Stop a running job using its unique job ID.
+  status      Query the status of a job using its unique job ID.
+  stream      Stream the output of a running job using its unique job ID.
+  run         Run a local process in specified namespaces and cgroup.
+
+Options:
+  --help      Show this help message and exit.
+  --cert      Specify the client certificate (and optionally key) file path for mTLS authentication.
+              Defaults to ~/.ssl/client.pem if not specified.
+
+Examples:
+  jobworker start --cmd "bash myscript.sh"                          Start a job using default certificate and key.
+  jobworker stop --job-id 12345 --cert /path/to/custom.pem          Stop a job using a specified PEM file.
+  jobworker status --job-id 12345                                   Query the status of a job using the default certificate and key.
+  jobworker stream --job-id 12345 --cert /path/to/custom.pem        Stream output of a job using a specified PEM file.
+  jobworker run --cmd "ls -l"                                       Run 'ls -l' in new PID, mount, and network namespaces with specified cgroup limits.
+
+Note:
+  'run' command requires root privileges to manage namespaces and cgroups.
+```
+
+### Security Considerations
+- **Authentication**: In this prototype, mTLS will be exclusively used for
```

Comment:
Can you add a bit more details on how we go from mTLS authentication to actually allowing/not allowing to fx stop job from another user.

Reply:
Will do.